### PR TITLE
Numpy.rff.rfft2 frontend function

### LINKED ISSUE #16357: numpy  rfft2
#1532

### PR DESCRIPTION
**Pull request by rajveer43**
Close #16357

### CODE REVIEW

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import ivy
 from ivy.functional.frontends.numpy.func_wrapper import to_ivy_arrays_and_back
-from ivy.func_wrapper import with_unsupported_dtypes
+from ivy.func_wrapper import with_unsupported_dtypes, with_supported_dtypes
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:3]: Don't know why you imported the `with_supported_dtypes` here since it is not being referenced to anywhere in the file?

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 2]
```diff
@@ -125,3 +125,16 @@ def rfftfreq(n, d=1.0):
...
+@to_ivy_arrays_and_back
+@with_unsupported_dtypes({"1.24.3 and above": ("float16",)}, "numpy")
+def rfft2(a, s=None, axes=(-2, -1), norm=None):
+    a = ivy.array(a, dtype=ivy.float64)
+    if norm is None:
+        norm = "backward"
+    # if s is None:
+    #     s = a.shape
+    # else:
+    s = list(s)
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:139]: What happens when `s` is None? The `s = list(s)` line will fail. Can you please add the logic for handling that case?

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 3]
```diff
@@ -125,3 +125,16 @@ def rfftfreq(n, d=1.0):
     N = n // 2 + 1
     results = ivy.arange(0, N, dtype=int)
     return results * val
+
+
+@to_ivy_arrays_and_back
+@with_unsupported_dtypes({"1.24.3 and above": ("float16",)}, "numpy")
+def rfft2(a, s=None, axes=(-2, -1), norm=None):
+    a = ivy.array(a, dtype=ivy.float64)
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:133]: Can you change this to `a = ivy.astype(a, "float64")`?

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 4]
```diff
@@ -125,3 +125,16 @@ def rfftfreq(n, d=1.0):
...
+    if norm is None:
+        norm = "backward"
+    # if s is None:
+    #     s = a.shape
+    # else:
+    s = list(s)
+    return ft.dft(a, axes=axes, inverse=False, onesided=True, dft_length=s, norm=norm)
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:140]: What is `ft.dft`?

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 5]
```diff
@@ -189,3 +189,30 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
         n=n,
         d=d,
     )
+
+
+@handle_frontend_test(
+    fn_tree="numpy.fft.rfft2",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("float_and_complex"),
+        min_num_dims=0 or None,
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:198]: I don't understand why we need to put `0 or None` here. Can you elaborate why you used this here?

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 6]
```diff
@@ -189,3 +189,30 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
...
+        max_num_dims=2,
+        force_tuple_axis=True
+    ),
+    norm=st.sampled_from(["backward", "ortho", "forward"]),
+    s=st.lists(st.integers()),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:203]: We also need to test for cases where `s` is None. You can use `st.one_of` to combine more than 1 strategies and pick one of them per each test run

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @rajveer43, thanks for opening the PR. Requested a couple of changes for you to address. Please link the issue/sub-task you are trying to work on by commenting `Close #<issue_number` on this PR. There also seem to be some linting/formatting issues with this PR. Please use `pre-commit` as specified in the documentation/website to auto-format your commits to the accepted standards. Can you also kindly run your tests locally using `pytest` and attach a screenshot of your logs here? Kindly ask for a re-review once you're done. Thanks! 😃

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py (hunk 5 above)
Comment (hmahmood24) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py]: Use `st.one_of` here as well

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 7]
```diff
@@ -125,3 +125,19 @@ def rfftfreq(n, d=1.0):
...
+    if isinstance(a, (list, ivy.array)):
+        a = ivy.array(a, "float64")
+        if a.size == 0:
+            return a
+    if norm is None:
+        norm = "backward"
+    if s is None:
+        s = a.shape
+    else:
+        s = tuple(s)
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:142]: Can you use a ternary operator to condense this condition so it looks more neat and readable. Something like:
s = tuple(s) if s else a.shape

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @rajveer43. Thanks for making the requested changes. I have reviewed and left some comments for you. Additionally, the CI is not picking up your tests. Can you run your tests locally using `pytest` and attach the test logs here? They are necessary to get this PR merged. Thanks! :)

**Comment by hmahmood24** [comment]
> 

@rajveer43 There is an issue with your installation of `ivy` I think. Did you install `ivy` using `pip` or by cloning it and installing it from source?

**Comment by hmahmood24** [comment]
> ```
> E:\MyJob\Ivy\ivy\ivy_dev\Scripts\python.exe "C:/Program Files/JetBrains/PyCharm 2023.1.2/plugins/python/helpers/pycharm/_jb_pytest_runner.py" --path ivy_dev/Lib/site-packages/ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py 
> Testing started at 2:09 PM ...
> Launching pytest with arguments ivy_dev/Lib/site-packages/ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py --no-header --no-summary -q in E:\MyJob\Ivy\ivy
> 
> Use Database in ReadOnly Mode with local caching !
> ============================= test session starts =============================
> collecting ... collected 0 items / 1 error
> 
> !!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
> ============================== 1 error in 0.40s ===============================
> 
> ivy_dev/Lib/site-packages/ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:None (ivy_dev/Lib/site-packages/ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py)
> ImportError while importing test module 'E:\MyJob\Ivy\ivy\ivy_dev\Lib\site-packages\ivy_tests\test_ivy\test_frontends\test_numpy\test_fft\test_discrete_fourier_transform.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> ivy_dev\lib\site-packages\_pytest\python.py:617: in _importtestmodule
>     mod = import_path(self.path, mode=importmode, root=self.config.rootpath)
> ivy_dev\lib\site-packages\_pytest\pathlib.py:564: in import_path
>     importlib.import_module(module_name)
> C:\Users\Lenovo\AppData\Local\Programs\Python\Python38\lib\importlib\__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> <frozen importlib._bootstrap>:1014: in _gcd_import
>     ???
> <frozen importlib._bootstrap>:991: in _find_and_load
>     ???
> <frozen importlib._bootstrap>:975: in _find_and_load_unlocked
>     ???
> <frozen importlib._bootstrap>:671: in _load_unlocked
>     ???
> ivy_dev\lib\site-packages\_pytest\assertion\rewrite.py:178: in exec_module
>     exec(co, module.__dict__)
> ivy_dev\Lib\site-packages\ivy_tests\test_ivy\test_frontends\test_numpy\test_fft\test_discrete_fourier_transform.py:7: in <module>
>     from ivy_tests.test_ivy.test_functional.test_experimental.test_nn.test_layers import (
> E   ModuleNotFoundError: No module named 'ivy_tests.test_ivy.test_functional.test_experimental'
> 
> Process finished with exit code 2
> ```
> 
> here are the logs

@rajveer43 There is an issue with your installation of `ivy` I think. Did you install `ivy` using `pip` or by cloning it and installing it from source?

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 8]
```diff
@@ -125,3 +125,13 @@ def rfftfreq(n, d=1.0):
     N = n // 2 + 1
     results = ivy.arange(0, N, dtype=int)
     return results * val
+
+
+@to_ivy_arrays_and_back
+@with_unsupported_dtypes({"1.24.3 and above": ("float16",)}, "numpy")
+def rfft2(a, s=None, axes=(-2, -1), norm=None):
+    a = ivy.array(a, "float64")
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:133]: Can you please change this to `a = ivy.astype(a, "float64")` ?

File: ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py [hunk 9]
```diff
@@ -125,3 +125,13 @@ def rfftfreq(n, d=1.0):
...
+    if norm is None:
+        norm = "backward"
+    s = a.shape if s is None else tuple(s)
+    return ivy.dft(a, axes=axes, inverse=False, onesided=True, dft_length=s, norm=norm)
```
Comment (hmahmood24) [inline ivy/functional/frontends/numpy/fft/discrete_fourier_transform.py:137]: `ivy.dft` has an `axis` kwarg that expects the value to be an `int`, not a sequence of `ints`. You will have to change your implementation to either use some other suitable function from `ivy` or implement this function as a composition of `ivy` functions if you can't find a suitable function in `ivy`.

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 10]
```diff
@@ -192,3 +192,33 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
         n=n,
         d=d,
     )
+
+
+@handle_frontend_test(
+    fn_tree="numpy.fft.rfft2",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("valid"),
+        min_num_dims=st.one_of(st.integers(min_value=0), st.none()),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:201]: `min_num_dims` is an argument to our helper function `dtype_values_axis` which expects this argument to be an integer just like you've passed in `max_num_dims`

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @rajveer43, requested changes on the PR. Additionally if you feel like you're stuck and don't know how to proceed further with this, please feel free to reach out to the team in the [numpy](https://discord.com/channels/799879767196958751/1028707030199980063) frontend channel. Thanks and good luck!

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 11]
```diff
@@ -192,3 +192,33 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
         n=n,
         d=d,
     )
+
+
+@handle_frontend_test(
+    fn_tree="numpy.fft.rfft2",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("valid"),
+        min_num_dims=st.integers(min_value=0),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:201]: Again, like I mentioned earlier, `min_num_dims` is supposed to be passed as an `int`, not a hypothesis strategy, similar to `max_num_dims`.

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 12]
```diff
@@ -192,3 +192,33 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
...
+        force_tuple_axis=True,
+    ),
+    norm=st.sampled_from(["backward", "ortho", "forward"]),
+    s=st.one_of(st.lists(st.integers()), st.none()),
+    axes=st.one_of(st.none(), st.lists(st.integers())),
+)
+def test_numpy_rfft2(
+    dtype_input_axis, norm, s, axes, frontend, test_flags, fn_tree, on_device
+):
+    input_dtype, x, axis = dtype_input_axis
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:212]: Use this `axis` to pass to the function instead.

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py [hunk 13]
```diff
@@ -192,3 +192,33 @@ def test_numpy_rfftfreq(n, sample_rate, frontend, test_flags, fn_tree, on_device
...
+    fn_tree="numpy.fft.rfft2",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("valid"),
+        min_num_dims=st.integers(min_value=0),
+        max_num_dims=2,
+        force_tuple_axis=True,
+    ),
+    norm=st.sampled_from(["backward", "ortho", "forward"]),
+    s=st.one_of(st.lists(st.integers()), st.none()),
+    axes=st.one_of(st.none(), st.lists(st.integers())),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_fft/test_discrete_fourier_transform.py:207]: This `axes` would not be needed since you are generating `axis` from `dtype_input_axis` above

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @rajveer43, requested changes for you. If the tests still fail for you, then they would probably be because of this line here: `s=st.one_of(st.lists(st.integers()), st.none())` cauisng `index out of range` errors for example. If that happens you'll probably need to implement a custom hypothesis strategy for the test that generates the valid input parameters. Try looking into the tests for `fft2` in other frontends to get an idea of how to generate valid data for this function since both the functions take in similar arguments to some extent. Thanks

**Comment by hmahmood24** [comment]
> Hi I did all the changes, I checked the failing test and found out that they are not related to rfft2, can you please check the changes.

@rajveer43 Can you please run your function locally? You've made an error that would break the test but I'm guessing you haven't tested your function locally at all?

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
Hey @rajveer43. The tests are still failing for all backends: https://github.com/unifyai/ivy/actions/runs/5465869385/jobs/9952016420?pr=17093#step:4:30. Did you run the test to verify if it gave the same result for all backends and for the native `numpy.fft.rfft2` function? You're generating `dim` separately whereas previously I suggested you to generate `axis` from `helpers.dtype_values_axis` by making `force_tuple_axis=True` and passing that to the `dim` argument. This will generate valid axis for you that won't fall outside the actual shape of `x` itself. Can you kindly test your implementation first and verify it works for all backends before re-requesting a review from me? Thanks
